src/util/pg.py

```diff
@@ -196,27 +196,41 @@ def explode_query_range(cr, query, table, alias=None, bucket_size=10000, prefix=
 
     alias = alias or table
 
-    cr.execute("SELECT min(id), max(id) FROM {}".format(table))
-    min_id, max_id = cr.fetchone()
-    if min_id is None:
+    cr.execute(
+        format_query(
+            cr,
+            """
+            SELECT min(id), max(id)
+              FROM (SELECT id,
+                           ROW_NUMBER() OVER (ORDER BY id) / %s
+                      FROM {table}
+                     ORDER BY id
+                   ) t(id, g)
+             GROUP BY g
+             ORDER BY g
+            """,
+            table=table,
+        ),
+        [bucket_size],
+    )
+    if not cr.rowcount:
         return []  # empty table
 
     if "{parallel_filter}" not in query:
         sep_kw = " AND " if re.search(r"\sWHERE\s", query, re.M | re.I) else " WHERE "
         query += sep_kw + "{parallel_filter}"
 
-    if ((max_id - min_id + 1) * 0.9) <= bucket_size:
-        # If there is less than `bucket_size` records (with a 10% tolerance), no need to explode the query.
-        # Force usage of `prefix` in the query to validate it correctness.
+    if cr.rowcount == 1:
```

Comment:
To keep the same "intended" behavior, we could check if there are only 2 buckets then count if the total number of records is at most 10% more than bucket_size. @KangOl should we do it?

Reply:
Now that we have exactly the correct number of records in the buckets, the 10% tolerance is not necessary.